BaseTools/Source/C/GenFw/GenFw.c

```diff
@@ -443,6 +443,64 @@ Routine Description:
   return STATUS_SUCCESS;
 }
 
+// MU_CHANGE begin
+BOOLEAN
+IsNxCompatCompliant (
+  EFI_IMAGE_OPTIONAL_HEADER_UNION  *PeHdr
+  )
+/*++
+
+Routine Description:
+
+  Checks if the Pe image is nxcompat. i.e. PE is 64bit, section alignment is
+  divisible by 4k, and all sections are not Write/Execute
+
+Arguments:
+
+  PeHdr      The Pe header
+
+Returns:
+
+  TRUE       The PE is nx compat compliant
+  FALSE      The PE is not nx compat compliant
+--*/
+{
+  EFI_IMAGE_SECTION_HEADER     *SectionHeader;
+  EFI_IMAGE_OPTIONAL_HEADER64  *Optional64;
+  UINT32                       Index;
+  UINT32                       Mask;
+
+  // Must have an optional header to perform verification
+  if (PeHdr->Pe32.FileHeader.SizeOfOptionalHeader == 0) {
+    return FALSE;
+  }
+
+  // Verify PE is 64 bit
+  if (!(PeHdr->Pe32.OptionalHeader.Magic == EFI_IMAGE_NT_OPTIONAL_HDR64_MAGIC)) {
+    return FALSE;
+  }
+
+  // Verify Section Alignment is divisible by 4K
+  Optional64 = (EFI_IMAGE_OPTIONAL_HEADER64 *)&PeHdr->Pe32.OptionalHeader;
+  if (!((Optional64->SectionAlignment % 0x1000) == 0)) {
+    return FALSE;
+  }
+  
+  // Verify sections are not Write & Execute
+  Mask = EFI_IMAGE_SCN_MEM_EXECUTE | EFI_IMAGE_SCN_MEM_WRITE;
+  SectionHeader = (EFI_IMAGE_SECTION_HEADER *) ((UINT8 *) &(PeHdr->Pe32.OptionalHeader) + PeHdr->Pe32.FileHeader.SizeOfOptionalHeader);

```

Comment:
Once you verify the magic value is `EFI_IMAGE_NT_OPTIONAL_HDR64_MAGIC` up above, I think you should start using the PE32+ format in the header union for consistency (even though most of the fields I see are in the common image header before the optional header).

Reply:
I have switched to using PeHdr->Pe32Plus once confirming HDR64_MAGIC